packages/near-sdk-js/src/utils.ts

```diff
@@ -147,6 +197,105 @@ export function deserialize(valueToDeserialize: Uint8Array): unknown {
   });
 }
 
+export function decodeObj2class(class_instance, obj) {
+  if (
+    typeof obj != "object" ||
+    class_instance.constructor.schema === undefined
+  ) {
+    return obj;
+  }
+  let key;
+  for (key in obj) {
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore
+    const value = obj[key];
+    if (typeof value == "object") {
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+      // @ts-ignore
+      const ty = class_instance.constructor.schema[key];
+      // eslint-disable-next-line no-prototype-builtins
+      if (ty !== undefined && ty.hasOwnProperty("map")) {
+        for (const mkey in value) {
+          if (ty["map"]["value"] === "string") {
+            class_instance[key][mkey] = value[mkey];
+          } else {
+            class_instance[key][mkey] = decodeObj2class(
+              new ty["map"]["value"](),
+              value[mkey]
+            );
+          }
+        }
+        // eslint-disable-next-line no-prototype-builtins
+      } else if (ty !== undefined && ty.hasOwnProperty("array")) {
+        for (const k in value) {
+          if (ty["array"]["value"] === "string") {
+            class_instance[key].push(value[k]);
+          } else {
+            class_instance[key].push(
+              decodeObj2class(new ty["array"]["value"](), value[k])
+            );
+          }
+        }
+        // eslint-disable-next-line no-prototype-builtins
+      } else if (ty !== undefined && ty.hasOwnProperty(UNORDERED_MAP_SCHE)) {
```

Comment:
Can we check if `ty` has `reconstructor` method, and call that instead of check `UNORDERED_MAP_SCHE`, etc.? If this is possible I feel it is more general and extensible, because assume user implement a new collection class, if he has `reconstruct` method implemented, check `reconstructor` approach will automatically work, but this way will need modify decodeObj2class.

On highlevel this is like: if `schema` exist, use `schema`. Otherwise try `reconstructor`. Else, simply `class_instance[key] = obj[key] `

Reply:
ok, I will try it

Reply:
resolve! I use this format in schema, a general `collection` type with `reconstructor` method:
```js
@NearBindgen({})
export class StatusDeserializeClass {
    static schema = {
        is_inited: "boolean",
        records: {map: { key: 'string', value: 'string' }},
        truck: Truck,
        messages: {array: {value: 'string'}},
        efficient_recordes: {collection: {reconstructor: UnorderedMap.reconstruct, value: 'string'}},
        nested_efficient_recordes: {collection: {reconstructor: UnorderedMap.reconstruct, value: { collection: {reconstructor: UnorderedMap.reconstruct, value: 'string'}}}},
        nested_lookup_recordes: {collection: {reconstructor: UnorderedMap.reconstruct, value: { collection: {reconstructor: LookupMap.reconstruct, value: 'string'}}}},
        vector_nested_group: {collection: {reconstructor: Vector.reconstruct, value: { collection: {reconstructor: LookupMap.reconstruct, value: 'string'}}}},
        lookup_nest_vec: {collection: {reconstructor: LookupMap.reconstruct, value: { collection: { reconstructor: Vector.reconstruct, value: 'string' }}}},
        unordered_set: {collection: {reconstructor: UnorderedSet.reconstruct, value: 'string'}},
        user_car_map: {collection: {reconstructor: UnorderedMap.reconstruct, value: Car }},
        big_num: 'bigint',
        date: 'date'
    };
    constructor() {
        this.is_inited = false;
        this.records = {};
        this.truck = new Truck();
        this.messages = [];
        // account_id -> message
        this.efficient_recordes = new UnorderedMap("a");
        // id -> account_id -> message
        this.nested_efficient_recordes = new UnorderedMap("b");
        // id -> account_id -> message
        this.nested_lookup_recordes = new UnorderedMap("c");
        // index -> account_id -> message
        this.vector_nested_group = new Vector("d");
        // account_id -> index -> message
        this.lookup_nest_vec = new LookupMap("e");
        this.unordered_set = new UnorderedSet("f");
        this.user_car_map = new UnorderedMap("g");
        this.big_num = 1n;
        this.date = new Date();
        this.message_without_schema_defined = "";
        this.number_without_schema_defined = 0;
        this.records_without_schema_defined = {};
    }
// ..
}
```

Reply:
This is challenging and your implementation is great! just one little catch, 
```
        nested_efficient_recordes: {collection: {reconstructor: UnorderedMap.reconstruct, value: { collection: {reconstructor: UnorderedMap.reconstruct, value: 'string'}}}},
```
API looks a little bit unfriendly, can we slightly generalize and improve the API, to be:

```
nested_efficient_recordes: {class: UnorderedMap, value: UnorderedMap} // value: UnorderedMap is a shortcut for value: {class: UnorderedMap, value:'string'}
```

sdk should first try if UnorderedMap has a `static schema` and use it to recursively decode. In this case, UnorderedMap doesn't. So sdk should next try call UnorderedMap.reconstruct.

Reply:
ok, I will try it